### DISCUSSION TITLE
page.search direction

### DISCUSSION THREAD

**Discussion by yonglee7015** [jsvine/pdfplumber#841] [Ask for help with specific PDFs]
is there any param can control the search direction in a page?  for example , from top to bottom or from left to right in a page.

**Comment by jsvine** [comment]
Currently, there is no such parameter. However, perhaps it helps if I clarify how search works:

- `pdfplumber` extracts the text from the page, and maps the extracted text to the original characters and their positions: https://github.com/jsvine/pdfplumber/blob/ee48b26099a614b9e97465963a5ff46aa2b04e46/pdfplumber/utils/text.py#L75-L88
- Then, `pdfplumber` uses Python's `re.finditer(pattern, text)` on that extracted text: https://github.com/jsvine/pdfplumber/blob/ee48b26099a614b9e97465963a5ff46aa2b04e46/pdfplumber/utils/text.py#L46-L63
- ... and returns a slightly processed version of the results of that call: https://github.com/jsvine/pdfplumber/blob/ee48b26099a614b9e97465963a5ff46aa2b04e46/pdfplumber/utils/text.py#L32-L44

Since this approach depends on [`re.finditer`](https://docs.python.org/3/library/re.html): If you wanted to use Python's regular expression tools to search bottom to top or left to right, how would you do it?

**Comment by yonglee7015** [comment]
re.finditer(pattern, string, flags=0)
Return an [iterator](https://docs.python.org/3/glossary.html#term-iterator) yielding [match objects](https://docs.python.org/3/library/re.html#match-objects) over all non-overlapping matches for the RE pattern in string. The string is scanned left-to-right, and matches are returned in the order found. Empty matches are included in the result.

 I just wondering if we can get the x,y position for each words . why we couldn't use it to indicate the direction from bottom to top or from right to left.

**Reply by jsvine** [reply]
Perhaps I have misunderstood the question. Are you just looking to get the _same_ search results as the default approach, but in a different order? If so, you can sort the result list by each item's `x0` and `top` properties. E.g.,:

```python
results = page.search("my phrase")
reordered = sorted(results, key=lambda x: (-x["top"], -x["x0"]))
```

Is that what you were looking for? If not, could you provide an example of the functionality you're seeking?

**Comment by yonglee7015** [comment]
aha,yes it is what I am looking for. Thanks for your kindly help.
